wordpress-mobile/WordPress-iOS
# MediaService: Cleanup part 2, cacheing, URLs, and tests.

This is technically another cleanup PR as it clears the path by changing the URL and file handling of actual Media files on disk via the local Media directory.

For the most part the "Media" directory is used as cache, but it's also kind of used as a scratch disk for uploads. While this Media directory should actually be within the application "Caches" directory, we'll keep it within "Documents" for now as I refresh the actual handling of exported assets on and off disk, in the next PRs.

Notable changes:

1. `clearUnusedFilesFromLocalMediaDirectory` and `clearCachedFilesFromLocalMediaDirectory ` were formerly seen independently as `cleanUnusedMediaFilesFromMediaCacheFolder` and `cleanMediaCacheFolder `: https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/Services/MediaService.h#L175 
   - Which now share `purgeLocalMediaFiles` as the actual implementation.
2. New tests under `Services/MediaServiceTests.swift` and `Models/MediaTests.swift`.
3. `AppSettingsViewController.swift` now handles the newly async nature of both calculating and clearing the local media directory, via new methods.
   - Also moves the `MediaSizingRow` out of `WPImmuTableRows.swift`
4. Refactors `Media` absolute URL handling to use `NSURL` and be relative to the `MediaService.localMediaDirectory()`.

To test:

1. Run all project tests.

To test:

1. Test that inserting and posting an image from the device library works.
2. Test that inserting and posting an video from the device library works.
3. Test that inserting and posting an image from the Media library works.
4. Test that inserting and posting a video from the Media library works.
5. Test that inserting and posting an image that is pasted into the editor works.
6. Test that inserting and posting a GIF from the device library works.
7. Test that inserting and posting a GIF from the Media library works.

To test:
1. Go to Me > App Settings.
2. Clear the cache.
3. Check that the UI was async, or at least it didn't noticeably lock.
4. Most of the cache should be cleared, some items may remain.
5. Background app.
6. Kill app.
7. Relaunch, but look for the log `"Media: removed \(removedCount) file(s) during cleanup."` in the console.
8. If no log appeared, switch blogs and repeat step 5-7, the log should show up when switching blogs or every so often.

Needs review: @frosty can I get your eyes on this one? Also, cc @SergioEstevao to keep an eye on the deeper implementation changes.

Note: Media in general is still a WIP, with plenty still yet to change. If something looks a bit funky with naming, or separation of concerns, bring it up and I'll keep it in mind for the next PRs.
